esdt-safe/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+#![no_std]
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+use elrond_wasm::{api::StorageReadApi, HexCallDataSerializer};
+
+const ESDT_SYSTEM_SC_ADDRESS_ARRAY: [u8; 32] = [
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0xff, 0xff,
+];
+
+const ESDT_BURN_STRING: &[u8] = b"ESDTBurn";
+
+#[derive(TopEncode, TopDecode, TypeAbi)]
+pub struct Transaction<BigUint: BigUintApi> {
+    from: Address,
+    to: Address,
+    token_identifier: TokenIdentifier,
+    amount: BigUint,
+}
+
+#[derive(TopEncode, TopDecode, TypeAbi, PartialEq)]
+pub enum TransactionStatus {
+    None,
+    Pending,
+    InProgress,
+    Executed,
+    Rejected,
+}
+
+#[elrond_wasm_derive::contract(EsdtSafeImpl)]
+pub trait EsdtSafe {
+    #[init]
+    fn init(&self, transaction_fee: BigUint, token_whitelist: &[TokenIdentifier]) {
+        self.set_transaction_fee(&transaction_fee);
+
+        for token in token_whitelist {
+            self.token_whitelist().insert(token.clone());
+        }
+    }
+
+    // endpoints - owner-only
+    // the owner will probably be a multisig SC
+
+    #[endpoint(setTransactionFee)]
+    fn set_transaction_fee_endpoint(&self, transaction_fee: BigUint) -> SCResult<()> {
+        only_owner!(self, "only owner may call this function");
+
+        self.set_transaction_fee(&transaction_fee);
+
+        Ok(())
+    }
+
+    #[endpoint(addTokenToWhitelist)]
+    fn add_token_to_whitelist(&self, token_identifier: TokenIdentifier) -> SCResult<()> {
+        only_owner!(self, "only owner may call this function");
+
+        self.token_whitelist().insert(token_identifier);
+
+        Ok(())
+    }
+
+    #[endpoint(removeTokenFromWhitelist)]
+    fn remove_token_from_whitelist(&self, token_identifier: TokenIdentifier) -> SCResult<()> {
+        only_owner!(self, "only owner may call this function");
+
+        self.token_whitelist().remove(&token_identifier);
+
+        Ok(())
+    }
+
+    #[endpoint(getNextPendingTransaction)]
+    fn get_next_pending_transaction(&self) -> SCResult<OptionalResult<Transaction<BigUint>>> {
```

Comment:
why do we need the pending transactions ?

Reply:
These are the transactions created by accounts that want to transfer to another chain. When they create a transaction, their tokens are locked inside the contract. Then, the owner can query the contract and get the pending transactions, execute them on the other chain, and then set their status to Executed (which burns the locked tokens) or Rejected (which will refund the tokens to the original sender).